File: android/sdk/src/main/kotlin/ai/edgevdb/TextChunker.kt
package ai.edgevdb

/**
 * Sliding-window text chunker.
 *
 * Splits a document into overlapping word-level chunks so that context
 * at chunk boundaries is not lost. BERT models handle up to 512 tokens
 * (~350–400 English words), so [chunkSize] ≤ 300 is recommended.
 *
 * @param chunkSize    Target words per chunk.
 * @param chunkOverlap Words of overlap between consecutive chunks.
 */
class TextChunker(
    private val chunkSize:    Int = 200,
    private val chunkOverlap: Int = 40
) {
    init {
        require(chunkSize > 0)         { "chunkSize must be > 0" }
        require(chunkOverlap >= 0)     { "chunkOverlap must be >= 0" }
        require(chunkOverlap < chunkSize) { "chunkOverlap must be < chunkSize" }
    }

    /**
     * Chunk [text] into [DocumentChunk]s with sliding window.
     *
     * @param text  The full document text.
     * @param docId Identifier for the parent document.
     * @param page  Starting page number (increment externally for paged docs).
     */
    fun chunk(text: String, docId: String, page: Int = 0): List<DocumentChunk> {
        val words  = text.split(Regex("\\s+")).filter { it.isNotBlank() }
        val chunks = mutableListOf<DocumentChunk>()
        var start  = 0
        var index  = 0

        while (start < words.size) {
            val end   = minOf(start + chunkSize, words.size)
            val chunk = words.subList(start, end).joinToString(" ")

            chunks += DocumentChunk(
                text       = chunk,
                docId      = docId,
                chunkIndex = index++,
                page       = page
            )

            start += chunkSize - chunkOverlap
        }

        return chunks
    }
}

File: ios/Sources/EdgeVDB/TextChunker.swift
import Foundation

/// TextChunker - Splits text into chunks for indexing
public struct TextChunker {
    /// Chunk size in words
    public let chunkSize: Int
    /// Overlap between chunks in words
    public let chunkOverlap: Int
    /// Minimum chunk length in characters
    public let minChunkLength: Int
    
    public init(chunkSize: Int = 200, chunkOverlap: Int = 40, minChunkLength: Int = 20) {
        self.chunkSize = chunkSize
        self.chunkOverlap = chunkOverlap
        self.minChunkLength = minChunkLength
    }
    
    /// Split text into chunks
    public func chunk(_ text: String) -> [DocumentChunk] {
        let words = text.components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
        var chunks: [DocumentChunk] = []
        var page = 0
        var i = 0
        
        while i < words.count {
            let end = min(i + chunkSize, words.count)
            let chunkWords = Array(words[i..<end])
            let chunkText = chunkWords.joined(separator: " ")
            
            // Skip trivially small final fragments
            if chunkText.count < minChunkLength {
                break
            }
            
            let chunk = DocumentChunk(
                text: chunkText,
                docId: "",
                page: page
            )
            chunks.append(chunk)
            
            page += 1
            i += chunkSize - chunkOverlap
        }
        
        return chunks
    }
    
    /// Split text into chunks with document ID
    public func chunk(_ text: String, docId: String) -> [DocumentChunk] {
        let chunks = chunk(text)
        return chunks.map { chunk in
            DocumentChunk(
                text: chunk.text,
                docId: docId,
                page: chunk.page
            )
        }
    }
}
